lib/store_spec.ml

```diff
@@ -24,39 +24,59 @@ let pp f = function
 
 type store = Store : (module S.STORE with type t = 'a) * 'a -> store
 
-let to_store mode = function
+let to_store rsync_mode = function
   | `Btrfs path ->
     Btrfs_store.create path >|= fun store ->
     Store ((module Btrfs_store), store)
   | `Zfs pool ->
     Zfs_store.create ~pool >|= fun store ->
     Store ((module Zfs_store), store)
   | `Rsync path ->
-    Rsync_store.create ~path ~mode () >|= fun store ->
+    Rsync_store.create ~path ~mode:rsync_mode () >|= fun store ->
     Store ((module Rsync_store), store)
 
-let cmdliner =
-  let open Cmdliner in
-  let store_t = Arg.conv (of_string, pp) in
-  let store =
-    Arg.required @@
-    Arg.opt Arg.(some store_t) None @@
-    Arg.info
-      ~doc:"$(b,btrfs:/path) or $(b,rsync:/path) or $(b,zfs:pool) for build cache."
-      ~docv:"STORE"
-      ["store"]
-  in
-  let rsync_mode =
-    let options =
-      [("copy", Rsync_store.Copy);
-       ("hardlink", Rsync_store.Hardlink);
-       ("hardlink_unsafe", Rsync_store.Hardlink_unsafe)]
-    in
-    Arg.value @@
-    Arg.opt (Arg.enum options) Rsync_store.Copy @@
-    Arg.info
-      ~doc:"$(b,copy) or $(b,hardlink), to optimize for speed or low disk usage."
-      ~docv:"RSYNC_MODE"
-      ["rsync-mode"]
+open Cmdliner
+
+let store_t = Arg.conv (of_string, pp)
+
+let store =
+  Arg.required @@
+  Arg.opt Arg.(some store_t) None @@
+  Arg.info
+    ~doc:"$(docv) must be one of $(b,btrfs:/path), $(b,rsync:/path) or $(b,zfs:pool) for the build cache."
+    ~docv:"STORE"
+    ["store"]
+
+let store' =
+  Arg.value @@
+  Arg.opt Arg.(some store_t) None @@
+  Arg.info
+    ~doc:"$(docv) must be one of $(b,btrfs:/path), $(b,rsync:/path) or $(b,zfs:pool) for the OBuilder cache."
+    ~docv:"STORE"
+    ["obuilder-store"]
+
+let rsync_mode =
+  let options =
+    [("copy", Rsync_store.Copy);
+     ("hardlink", Rsync_store.Hardlink);
+     ("hardlink_unsafe", Rsync_store.Hardlink_unsafe)]
   in
+  Arg.value @@
+  Arg.opt (Arg.enum options) Rsync_store.Copy @@
+  Arg.info
+    ~doc:(Printf.sprintf "Optimize for speed or low disk usage. $(docv) must be one of %s."
+            (Arg.doc_alts_enum options))
+    ~docv:"RSYNC_MODE"
+    ["rsync-mode"]
+
+(** A Cmdliner term where the store is required. *)
+let cmdliner =
   Term.(const to_store $ rsync_mode $ store)
+
+(** A Cmdliner term where the store is optional. *)
+let cmdliner' =
+  let make rsync_mode = function
+    | None -> None
+    | Some store -> Some (to_store rsync_mode store)
```

Comment:
This `to_store` function should be without side-effects, eg just reading the cli argument and then later in the program we run the `create` associated with each store.

Reply:
One idea would be to have a new type describing the configuration of the store, like a record type created from the cmdliner options, and use it to create the store later on. I feel however that it's a bit invasive for this PR only.
In my main Windows branch, I've changed the function to return this `Store_spec.store Lwt.t option` which effectively delays the side-effects until the option is unwrapped and the promise executed, that is IMO easier.